File: src/main/scala/im/tox/toktok/BundleKey.scala
package im.tox.toktok

import android.os.Bundle
import im.tox.toktok.TypedBundleKey.{ IntKey, StringKey }

import scala.annotation.varargs

object BundleKey {
  val colorPrimary = IntKey("colorPrimary")
  val colorPrimaryStatus = IntKey("colorPrimaryStatus")
  val contactColorPrimary = IntKey("contactColorPrimary")
  val contactColorStatus = IntKey("contactColorStatus")
  val contactName = StringKey("contactName")
  val contactPhotoReference = IntKey("contactPhotoReference")
  val imgResource = IntKey("imgResource")
  val messageTitle = StringKey("messageTitle")
  val messageType = IntKey("messageType")
}

object TypedBundleKey {
  sealed abstract class KeyValue[A] {
    def put(bundle: Bundle): Unit
  }

  final case class StringKeyValue(id: String, value: String) extends KeyValue[String] {
    def put(bundle: Bundle): Unit = bundle.putString(id, value)
  }
  final case class IntKeyValue(id: String, value: Int) extends KeyValue[Int] {
    def put(bundle: Bundle): Unit = bundle.putInt(id, value)
  }

  final case class StringKey(id: String) {
    def ->(value: String): StringKeyValue = StringKeyValue(id, value)
  }
  final case class IntKey(id: String) {
    def ->(value: Int): IntKeyValue = IntKeyValue(id, value)
  }

  implicit class TypedBundle(val bundle: Bundle) extends AnyVal {
    def apply(key: StringKey): String = bundle.getString(key.id)
    def apply(key: IntKey): Int = bundle.getInt(key.id)
  }

  @varargs
  def SBundle(keyValues: KeyValue[_]*): Bundle = {
    val bundle = new Bundle
    keyValues.foreach(_.put(bundle))
    bundle
  }
}

File: src/main/java/im/tox/toktok/app/BundleKey.java
package im.tox.toktok.app;

import android.support.annotation.NonNull;

import im.tox.toktok.app.TypedBundleKey.IntKey;
import im.tox.toktok.app.TypedBundleKey.StringKey;

public class BundleKey {
    public static final IntKey colorPrimary = new IntKey("colorPrimary");

    @NonNull
    public static final IntKey colorPrimary() {
        return colorPrimary;
    }

    public static final IntKey colorPrimaryStatus = new IntKey("colorPrimaryStatus");

    @NonNull
    public static final IntKey colorPrimaryStatus() {
        return colorPrimaryStatus;
    }

    public static final IntKey contactColorPrimary = new IntKey("contactColorPrimary");

    @NonNull
    public static final IntKey contactColorPrimary() {
        return contactColorPrimary;
    }

    public static final IntKey contactColorStatus = new IntKey("contactColorStatus");

    @NonNull
    public static final IntKey contactColorStatus() {
        return contactColorStatus;
    }

    public static final StringKey contactName = new StringKey("contactName");

    @NonNull
    public static final StringKey contactName() {
        return contactName;
    }

    public static final IntKey contactPhotoReference = new IntKey("contactPhotoReference");

    @NonNull
    public static final IntKey contactPhotoReference() {
        return contactPhotoReference;
    }

    public static final IntKey imgResource = new IntKey("imgResource");

    @NonNull
    public static final IntKey imgResource() {
        return imgResource;
    }

    public static final StringKey messageTitle = new StringKey("messageTitle");

    @NonNull
    public static final StringKey messageTitle() {
        return messageTitle;
    }

    public static final IntKey messageType = new IntKey("messageType");

    @NonNull
    public static final IntKey messageType() {
        return messageType;
    }
}
